### PR TITLE
Add support for optgroup-tags for select-fields

### LINKED ISSUE #1776: Form: attribute type not allowed in select tag
Currently a dropdown created with DokuWiki's `Form` class has invalid html: A `<select>` tag must not have a `type` attribute. You can verify this error by pasting the code below into a validator like https://validator.w3.org/nu/#textarea

```html
<!DOCTYPE html>
<html>
<head>
<title>Title</title>
</head>
<body>
<select name="name" type="dropdown"><option selected="selected" value="1">1</option><option value="2">2</option></select>
</body>
</html>
```

### PR DESCRIPTION
**Pull request by micgro42**
In more complex selects we may want to group options by some criteria. HTML has the `<optgroup>`-tag for that purpose.

However in order to not duplicate code, I have chosen to move the handling of options to the `OptGroup`-class.

ToDo
-------
- [x] Tests

Known Issues:
------------------
* We no longer ensure that there are no duplicate options. Whereas before there could be no two options with the same value, because every option's value was a key in an associative array, now there can be two options with the same value within different optgroups
* This has also the consequence that there might be more than two options with the `selected`-attribute.

This also fixes #1776

### CODE REVIEW

**Comment by splitbrain** [comment]
haven't looked at the code, yet. but from your description I have a question. what happens if two options are selected at the same time? I guess the browser will only select one. but is it even allowed? can we prevent it?

**Comment by micgro42** [comment]
- [ ] If it is not a multi-select: set value only for first occurrence of value

File: inc/Form/DropdownElement.php [hunk 1]
```diff
@@ -91,18 +123,57 @@ public function attr($name, $value = null) {
...
+        $value_exists = $this->setValueInOptGroups($value);
+
+        if($value_exists) {
             $this->value = $value;
         } else {
             // unknown value set, select first option instead
-            $keys = array_keys($this->options);
-            $this->value = (string) array_shift($keys);
+            $this->value = $this->getFirstOption();
+            $this->setValueInOptGroups($this->value);
```
Comment (splitbrain) [inline inc/Form/DropdownElement.php]: how are we handling multi selects? it seems this only works with a single value?

File: inc/Form/DropdownElement.php [hunk 2]
```diff
@@ -111,15 +182,8 @@ protected function mainElementHTML() {
...
-            $selected = ($key == $this->value) ? ' selected="selected"' : '';
-            $attrs = '';
-            if (is_array($val['attrs'])) {
-                array_walk($val['attrs'],function (&$aval, $akey){$aval = hsc($akey).'="'.hsc($aval).'"';});
-                $attrs = join(' ', $val['attrs']);
-            }
-            $html .= '<option' . $selected . ' value="' . hsc($key) . '" '.$attrs.'>' . hsc($val['label']) . '</option>';
-        }
+        $html .= $this->options->toHTML();
+        $html = array_reduce($this->optGroups, function($html, OptGroup $optGroup) {return $html . $optGroup->toHTML();}, $html);
```
Comment (splitbrain) [inline inc/Form/DropdownElement.php]: I think it's a bit weird that not all options are handled through the optgroups in `$this->optGroups` but that there are two places. I would expect the "unlabeled" optgroup to be in that array as well.

File: inc/Form/DropdownElement.php [hunk 3]
```diff
@@ -21,7 +25,46 @@ class DropdownElement extends InputElement {
...
+        if($optGroups === null) {
+            return $this->optGroups;
+        }
+        if (!is_array($optGroups)) {
+            throw new \InvalidArgumentException(hsc('Argument must be an associative array of label => [options]!'));
+        }
+        $this->optGroups = array();
+        foreach ($optGroups as $label => $options) {
+            $this->addOptGroup($label, $options);
+        }
```
Comment (micgro42) [inline inc/Form/DropdownElement.php]: Fixed in 2bd1d2c

File: inc/Form/DropdownElement.php (hunk 1 above)
Comment (micgro42) [inline inc/Form/DropdownElement.php]: We are not, we are throwing an exception when setting the `multiple`-attribute. I removed my erroneous handling of multi selects in  45082b9.
We should consider a new PR for adding multi-select-functionality.

**Pull request merged**